hlf/scala/general-communication.md

```diff
@@ -1,79 +1,105 @@
+
+protected val caURL: String = "172.17.0.3:30906" // address of the CA-server.
+
+protected val username: String = "TestUser123" // this should in most cases be the name of the .id file in your wallet directory.
+protected val password: String = "Test123" // a password used to register a user and receive/set a certificate for said user when enrolling.
+protected val organizationId: String = "org1MSP" // the id of the organization the user belongs to.
+
+protected val organizationName: String = "org1" // the name of the organization the user belongs to.
+
+```
+
+## Regular Communication
+All chaincode-transactions are per default provided via a Connection-Pattern.
+- Here the USER has to provide their credentials to authenticate themselves when initiating a connection.
+    This is usually done by providing a wallet (walletPath) containing the Keys of the User ( in a file "userName.id")
+```scala
+val certificateConnection: ConnectionCertificateTrait = 
+    ConnectionCertificate(userName, this.channel, this.chaincode, this.walletPath, this.networkDescriptionPath)
+```
+- Subsequent proposals are then signed with the credentials belonging to the USER initiating the connection.
+    This simplifies the API-handling to method-calls.
+```scala
+val certificate: String = Source.fromURL(getClass.getResource("/testid.csr")).mkString
+certificateConnection.addCertificate(enrollmentId, certificate)
+```
+
+## THIRD-PARTY-Signed Communication
+All chaincode-transactions are furthermore provided as an "unsigned" proposal as well.
+
+- Here some sort of ADMIN has to provide their credentials [..] when initiating a connection.
+```scala
+val certificateConnection: ConnectionCertificateTrait = 
+    ConnectionCertificate(adminUserName, this.channel, this.chaincode, this.adminWalletPath, this.networkDescriptionPath)
+```
+The ADMIN can then 
+- request such an "unsigned" proposal
+```scala
+val proposal: String = certificateConnection.getProposalAddCertificate(enrollmentId, certificate)
+```
+
+> Note!
+> The proposal is a simple Protobuf-object.
+
+ **_IMPORTANT: When asking for a proposal, the credentials used for setting up the initial connection are used to approve the transaction as well (in this case as the ADMIN)_**
+- pass it on to the USER that wants to sign it
+- have the USER create his signature for the "unsigned" proposal
+- receive the signature from the USER (together with the original proposal)
+
+- request the unsigned transaction by submitting the signed proposal
+
+```scala
+val (unsignedTransaction, proposalTransactionId) = certificateConnection.getUnsignedTransaction(proposalBytes: Array[Byte], signature: Array[Byte])
+```
+> Note!
+> This process **does** communicate with the HLF Network. 
+> It is very slow.
+
+> Note!
+> The unsigned transaction returned is a protobuf-object as well.
+
+> Note!
+> The proposalTransactionId has to be passed to the submitTransaction-call.
+> We do not yet know if we can omit it.

```

Comment:
From what I found, the transaction consists of a header and data. The header consists of channelHeader and signatureHeader. 
Inside the channelHeader, you can finde the txId, which is the same id as the one in the proposal.

Reply:
@NikoBergemann Therefore we can omit the transactionId, can we not?

Reply:
yes, please update this branch from develop (take source) after PR https://github.com/upb-uc4/api/pull/92 has been completed